### PR TITLE
Add enum.StrEnum and enum.IntEnum support for yt_dataclasses

### PR DESCRIPTION
**Pull request by thenno**
related to https://github.com/ytsaurus/ytsaurus/issues/195

### CODE REVIEW

File: yt/python/yt/wrapper/schema/types.py [hunk 1]
```diff
@@ -171,34 +200,33 @@ def _get_time_types():
 
 
 def _get_py_time_types():
-    if hasattr(_get_py_time_types, "_info"):
-        return _get_py_time_types._info
-    check_schema_module_available()
```
Comment (thenno) [inline yt/python/yt/wrapper/schema/types.py:176]: Maybe I've recklessly deleted this code, but it looks like it doesn't depend on python's version and skiff_bindings.

**Review by denvr — CHANGES_REQUESTED** [review]
Changes Requested

File: yt/python/yt/wrapper/schema/types.py (hunk 1 above)
Comment (thenno) [reply yt/python/yt/wrapper/schema/types.py:176]: Yes, it's common behavior, but that looks kinda strange.

I expected `check_schema_module_available` check in the top-level functions/objects, who really need skiff to work.

But I agree with you, it's not a good idea to change this only in one function.

File: yt/python/yt/wrapper/schema/types.py [hunk 2]
```diff
@@ -109,6 +112,32 @@ def create_annotated_type(py_type, ti_type, to_yt_type=None, from_yt_type=None):
     return Annotated[py_type, Annotation(ti_type, to_yt_type=to_yt_type, from_yt_type=from_yt_type)]
 
 
+def yt_enum(ti_type, to_yt_type=None, from_yt_type=None):
```
[the remark this answers is not in the document]
Comment (thenno) [reply yt/python/yt/wrapper/schema/types.py]: In my opinion, we shouldn't try to guess the type at runtime by content.
If we guess the type the user may get unpleasant effects, for example:

```
import enum

class C(enum.IntEnum):
    a = 1

# will be mapped to Int8

# after some time user can add another value

class C(enum.IntEnum):
    a = 1
    b = 214748364700

# now it maps to UInt64 implicitly
```

On the one hand we can try to guess the type by base type, but on the other hand we have to set the most general data type (for example Int64 for int), and this provokes suboptimal user behavior (nobody like to override defaults:) ).

File: yt/python/yt/wrapper/schema/types.py [hunk 3]
```diff
@@ -109,6 +112,32 @@ def create_annotated_type(py_type, ti_type, to_yt_type=None, from_yt_type=None):
...
+            c: List[CustomEnum]
+    """
+    def _wrapper(py_type):
+        return create_annotated_type(
+            py_type,
+            ti_type=ti_type,
+            to_yt_type=to_yt_type,
+            from_yt_type=from_yt_type,
+        )
+    return _wrapper
```
[the remark this answers is not in the document]
Comment (thenno) [reply yt/python/yt/wrapper/schema/types.py]: https://docs.python.org/3/library/typing.html#typing.Annotated -> typing._AnnotatedAlias

But the approach to the solution has been changed, please see the last comment.

File: yt/python/yt/wrapper/tests/typed_api/test_typed_api.py [hunk 4]
```diff
@@ -1340,6 +1342,69 @@ class RowWithTuple:
...
+    def test_enum(self):
+        @yt.yt_enum(ti.Int32)
+        class CustomIntEnum(enum.IntEnum):
+            A = 1
+            B = 2
+            C = 3
+            D = 4
+
+        @yt.yt_enum(ti.Utf8)
+        class CustomStrEnum(enum.StrEnum):
```
[the remark this answers is not in the document]
Comment (thenno) [reply yt/python/yt/wrapper/tests/typed_api/test_typed_api.py]: StrEnums was added in python3.11 -> so it should work.

Could you please clarify why you're asking this question? Opensource/closedsource ya make uses python 3.12.3, this file already contains new futures from 3.10 (new Union syntax). I can't see any potential problems here.

File: yt/python/yt/wrapper/tests/typed_api/test_typed_api.py [hunk 5]
```diff
@@ -1340,6 +1342,69 @@ class RowWithTuple:
...
         assert row.tuple_int_bytes == (1, b"\x01",)
 
+    @authors("thenno")
+    def test_enum(self):
+        @yt.yt_enum(ti.Int32)
+        class CustomIntEnum(enum.IntEnum):
+            A = 1
+            B = 2
+            C = 3
+            D = 4
```
[the remark this answers is not in the document]
Comment (thenno) [reply yt/python/yt/wrapper/tests/typed_api/test_typed_api.py]: It's a very good point.

I've checked only my usual use of enums, but annotated enum doesn't support all enum's magic.

For example, it damage iteration by enum-class: https://github.com/python/typing/discussions/1440
There is the same problem with dict-like access to the class. The worst thing is that it depends on the python version.

Thus my bold idea doesn't really work -> I propose to separate the annotation of the type and its actual use.

File: yt/python/yt/wrapper/schema/types.py [hunk 6]
```diff
@@ -109,6 +112,61 @@ def create_annotated_type(py_type, ti_type, to_yt_type=None, from_yt_type=None):
     return Annotated[py_type, Annotation(ti_type, to_yt_type=to_yt_type, from_yt_type=from_yt_type)]
 
 
+def create_yt_enum(py_type, ti_type, to_yt_type=None, from_yt_type=None):
```
Comment (denvr) [inline yt/python/yt/wrapper/schema/types.py:115]: Type annotations will make work easier (it not so easy to understand what `ti_type` is)

File: yt/python/yt/wrapper/tests/typed_api/test_typed_api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (denvr) [reply yt/python/yt/wrapper/tests/typed_api/test_typed_api.py]: My bad. I mean "below python 3.11". The closedsource `ya make` uses python 3.12, but the code should still support python 3.8+.
It turns out the tests were ahead of the code.

File: yt/python/yt/wrapper/schema/types.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (denvr) [reply yt/python/yt/wrapper/schema/types.py]: Of course defaults assumes wider types. As wide as regular integer field, without specific type annonations.

**Review by denvr — CHANGES_REQUESTED** [review]
Small notes in the tests, type hints

**Comment by denvr** [comment]
Shipped!

**Pull request closed without merging**